crypto/cmp/cmp_protect.c

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright 2007-2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright Nokia 2007-2019
+ * Copyright Siemens AG 2015-2019
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "cmp_local.h"
+
+/* explicit #includes not strictly needed since implied by the above: */
+#include <openssl/asn1t.h>
+#include <openssl/cmp.h>
+#include <openssl/crmf.h>
+#include <openssl/err.h>
+#include <openssl/x509.h>
+
+/*
+ * This function is also used for verification from cmp_vfy.
+ *
+ * Calculate protection for given PKImessage utilizing the given credentials
+ * and the algorithm parameters set inside the message header's protectionAlg.
+ *
+ * Either secret or pkey must be set, the other must be NULL. Attempts doing
+ * PBMAC in case 'secret' is set and signature if 'pkey' is set - but will only
+ * do the protection already marked in msg->header->protectionAlg.
+ *
+ * returns ptr to ASN1_BIT_STRING containing protection on success, else NULL
+ */
+ASN1_BIT_STRING *ossl_cmp_calc_protection(const OSSL_CMP_MSG *msg,
+                                          const ASN1_OCTET_STRING *secret,
+                                          EVP_PKEY *pkey)
+{
+    ASN1_BIT_STRING *prot = NULL;
+    CMP_PROTECTEDPART prot_part;
+    const ASN1_OBJECT *algorOID = NULL;
+    int l;
+    size_t prot_part_der_len;
+    unsigned char *prot_part_der = NULL;
+    size_t sig_len;
+    unsigned char *protection = NULL;
+    const void *ppval = NULL;
+    int pptype = 0;
+    OSSL_CRMF_PBMPARAMETER *pbm = NULL;
+    ASN1_STRING *pbm_str = NULL;
+    const unsigned char *pbm_str_uc = NULL;
+    EVP_MD_CTX *evp_ctx = NULL;
+    int md_NID;
+    const EVP_MD *md = NULL;
+
+    if (!ossl_assert(msg != NULL))
+        return NULL;
+
+    /* construct data to be signed */
+    prot_part.header = msg->header;
+    prot_part.body = msg->body;
+
+    l = i2d_CMP_PROTECTEDPART(&prot_part, &prot_part_der);
+    if (l < 0 || prot_part_der == NULL) {
```

Comment:
personally, I don't like variables named l because they look like I or 1 dependent on the display.

Reply:
in particular, when I read this if above, it looks like if (1 < 0 || ...) I have to read it twice.

Reply:
You are right. I renamed it to `len`.